Google Storage & Pub-Sub: Retry google token requests on 5xx

francisdb opened the pull request:
Fixes #2080


Some notes
* This adds 2s to the test time because the first backoff according to the google GCS SLA should be 1s
* I have not fully tested the negative case where a total 32 sec backoff is required
* Is this `RuntimeException` ok?

francisdb commented:
Would be great if this could make it to `2.0.0-M3`

ennru commented on google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/impl/GoogleSession.scala:
```diff
@@ -5,7 +5,7 @@
 
 import akka.annotation.InternalApi
-import akka.stream.Materializer
+import akka.stream.ActorMaterializer
```
Why this `ActorMaterializer`?

francisdb replied:
Removed where not needed

ennru commented on google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/impl/GoogleRetry.scala:
```diff
@@ -0,0 +1,54 @@
+        minBackoff = 1.second,
+        maxBackoff = 32.seconds,
+        randomFactor = 0
+      ) { () =>
```
I wonder if this could be solved more idiomatically with a `RetryFlow` (which would need be copied in for Akka 2.5 support).

francisdb replied:
I did not write this part, just moved it over here for reuse.

francisdb replied:
anyway, looking into this

francisdb replied:
Also synced both implementations, their files are almost identical

francisdb replied:
@ennru better like this?

francisdb commented:
both projects have the same tests but the storage one logs this during testing (test does not fail):

```
[ERROR] [02/14/2020 13:55:51.422] [default-akka.actor.default-dispatcher-5] [akka://default/system/StreamSupervisor-0/flow-2-0-headSink] Error during postStop in [akka.stream.alpakka.googlecloud.storage.impl.backport.RetryFlowCoordinator@7e6d0375]: null
java.lang.NullPointerException
	at akka.stream.stage.GraphStageLogic.afterPostStop(GraphStage.scala:1332)
	at akka.stream.stage.TimerGraphStageLogic.afterPostStop(GraphStage.scala:1608)
	at akka.stream.impl.fusing.GraphInterpreter.finalizeStage(GraphInterpreter.scala:580)
	at akka.stream.impl.fusing.GraphInterpreter.afterStageHasRun(GraphInterpreter.scala:560)
	at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:424)
	at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:606)
	at akka.stream.impl.fusing.GraphInterpreterShell$AsyncInput.execute(ActorGraphInterpreter.scala:485)
	at akka.stream.impl.fusing.GraphInterpreterShell.processEvent(ActorGraphInterpreter.scala:581)
	at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:749)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:764)
	at akka.actor.Actor.aroundReceive(Actor.scala:539)
	at akka.actor.Actor.aroundReceive$(Actor.scala:537)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:671)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:612)
  | => aat akka.actor.ActorCell.invoke(ActorCell.scala:581)okenApiSpec 2s
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:268)
	at akka.dispatch.Mailbox.run(Mailbox.scala:229)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:241)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

francisdb commented:
> Once we the solution is OK, we'd want to propagate it to Alpakka Google Cloud Storage.

This PR is for both?

ennru commented:
The `RetryFlow` shines here ;-)
Why the switch to `ActorMaterializer`? It should be just `Materializer` if there is no special need.

francisdb commented:
@ennru I removed all those `ActorMaterializer`s that I added before, were indeed not needed. The reason why `ActorMaterializer` is used in some places is because we need an `ActorSystem` to use `Http()`.

ennru commented on google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/impl/GoogleTokenApi.scala:
```diff
@@ -51,15 +51,30 @@ private[googlecloud] class GoogleTokenApi(http: => HttpExt) {
+        }
+      case other => Future.successful(other)
+    }
+  }
```
This could go into the flow in `GoogleRetry`.

ennru commented on google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/impl/GoogleRetry.scala:
```diff
@@ -0,0 +1,52 @@
+
+  // Exponential backoff as specified in the GCS SLA: https://cloud.google.com/storage/sla
+  // 1 initial attempt, plus 2^5 exponential requests to get to 32 seconds
+
```
This could be API docs.

ennru commented:
Thank you @francisdb, a great improvement!
